hashicorp/waypoint
# Add `UI_ListProjects`

## Why the change?

For the “destroy project” UI we want to show a “destroying” status on appropriate items in the project list page. The existing `ListProjects` method only gives us a list of `Ref.Project` which doesn’t include the status, so we’re proposing this new method which includes the stuff we need.

## What’s the plan?

- [x] Write basic handler test
- [x] Add `UI_ListProjects` to `server.proto` and `gateway.yml`
- [x] `make gen/server`
- [x] Add `ProjectListFull` (or something like that) to the state layer
- [x] Implement `boltdbstate.ProjectCount`
- [x] Implement `UI_ListProjects` handler
- [x] Add API validations
- [x] Decide whether to attempt or defer pagination for boltdb (outcome: defer)

## What does it look like?

<img width="792" alt="CleanShot 2023-02-03 at 14 23 40@2x" src="https://user-images.githubusercontent.com/34030/216614177-5f0345b9-a917-4c3e-b876-a28140947f26.png">

## How do I test it?

1. Check out the branch
2. `make bin`
3. Boot the server and connect to it
4. Init some projects
5. From the waypoint directory, run `./contrib/waypoint-grpc/waypoint-grpc.sh UI_ListProjects`
6. Verify all non-empty project-fields are present
7. Ignore pagination, we’ll come back to this later
